src/sources/youtube.rs

```diff
@@ -0,0 +1,173 @@
+            });
+
+            Some(lines.collect())
+        } else {
+            None
+        }
+    }
+}
+
+struct YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Sync,
+{
+    uri: P,
+}
+
+#[async_trait]
+impl<P> Restart for YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Clone + Sync,
+{
+    async fn call_restart(&mut self, time: Option<Duration>) -> SongbirdResult<Input> {
+        let (yt, metadata) = ytdl(self.uri.as_ref()).await?;
+
+        if let Some(time) = time {
+            let ts = format!("{:.3}", time.as_secs_f64());
+            ffmpeg_yt(yt, metadata, &["-ss", &ts]).await
+        } else {
+            ffmpeg_yt(yt, metadata, &[]).await
+        }
+    }
+
+    async fn lazy_init(&mut self) -> SongbirdResult<(Option<Metadata>, Codec, Container)> {
+        _ytdl_metadata(self.uri.as_ref())
+            .await
+            .map(|m| (Some(m), Codec::FloatPcm, Container::Raw))
+    }
+}
+
+async fn ytdl(uri: &str) -> Result<(Child, Metadata), SongbirdError> {
+    let ytdl_args = [
+        "--print-json",
+        "-f",
+        "bestaudio",
+        "-R",
+        "infinite",
+        "--no-playlist",
+        "--ignore-config",
+        "--no-warnings",
+        uri,
+        "-o",
+        "-",
+    ];
+
+    let mut yt = Command::new("yt-dlp")
+        .args(&ytdl_args)
+        .stdin(Stdio::null())
+        .stderr(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    // This rigmarole is required due to the inner synchronous reading context.
+    let stderr = yt.stderr.take();
+    let (returned_stderr, value) = task::spawn_blocking(move || {
+        let mut s = stderr.unwrap();
+        let out: SongbirdResult<Value> = {
+            let mut o_vec = vec![];
+            let mut serde_read = BufReader::new(s.by_ref());
+            // Newline...
+            if let Ok(len) = serde_read.read_until(0xA, &mut o_vec) {
+                serde_json::from_slice(&o_vec[..len]).map_err(|err| SongbirdError::Json {
+                    error: err,
+                    parsed_text: std::str::from_utf8(&o_vec).unwrap_or_default().to_string(),
+                })
+            } else {
+                Result::Err(SongbirdError::Metadata)
+            }
+        };
+
+        (s, out)
+    })
+    .await
+    .map_err(|_| SongbirdError::Metadata)?;
+
+    let metadata = Metadata::from_ytdl_output(value?);
+    yt.stderr = Some(returned_stderr);
+
+    Ok((yt, metadata))
+}
+
+async fn _ytdl_metadata(uri: &str) -> SongbirdResult<Metadata> {
+    let ytdl_args = [
+        "-j",
+        "-f",
+        "bestaudio",
+        "-R",
+        "infinite",
+        "--no-playlist",
+        "--ignore-config",
+        "--no-warnings",
+        uri,
+        "-o",
+        "-",
+    ];
+
+    let youtube_dl_output = TokioCommand::new("yt-dlp")
+        .args(&ytdl_args)
+        .stdin(Stdio::null())
+        .output()
+        .await?;
```

Comment:
Why do we use `TokioCommand` here and `Command` in line 102?

Reply:
Not really sure, the tokio variant is just an async implementation of Command but we await regardless.
Changed to `std::process::Command` and seems to run just fine.